### DISCUSSION TITLE
Query multiple JSON arrays of objects

### DISCUSSION THREAD

**Discussion by logikaljay** [supabase/supabase#3540] [Questions] [answered]
Hey Team,

Happy Friday (unless its not friday when you read this).

I have a table of email metadata:

| column | type | description |
| ------- | ---- | ------------ |
| id | uuid | primary key of table |
| uid | integer | unique id of email for the account and mailbox |
| mailbox | string | name of the mailbox the email belongs too, EG: Inbox, Sent Items, Junk |
| account_id | uuid | the email account this message belongs too |
| from | jsonb | the senders email address, eg: { name: "logikal", email: "logikal@gmail.com" }
| to | jsonb | an array of receivers email addresses eg: [{ name: "jim", email: "contact@jamesdraws.com" }]
| cc | jsonb | an array of cc email addresses eg: [{ name: "bob", email: "robert@officalcompanyname.com" }]
| message_id | string | unique identifier for the message for the email server |
| created_at | timestamp | the time the message was created on the server |

I am trying to get all messages that **involve** an email address, in a single query (so that I can paginate easily).

I am able to get all messages that match a `from` address:
```js
let { data: messages, error: findMessagesError } = await supabase
  .from('email_account_messages')
  .select('*')
  .contains('from', JSON.stringify({ address: "contact@jamesdraws.com" }))
```

I am able to get all messages that contain a `to` address:
```js
let { data: messages, error: findMessagesError } = await supabase
  .from('email_account_messages')
  .select('*')
  .contains('to', JSON.stringify([{ address: 'contact@jamesdraws.com' }])) // notice the array of addresses
```

I am able to get all messages that contain a `cc` address:
```js
let { data: messages, error: findMessagesError } = await supabase
  .from('email_account_messages')
  .select('*')
  .contains('cc', JSON.stringify([{ address: 'contact@jamesdraws.com' }])) // notice the array of addresses
```

However, I am not able to combine these queries to get any message that is sent to, sent from, or cc'd to a specific email address.

I have looked at the following discussions, which is how I got this far, but neither of them are checking multiple database columns for a single value.

* https://github.com/supabase/supabase/discussions/2387
* https://github.com/supabase/supabase/discussions/2102

Is there a way to achieve this without writing a function / stored procedure? (I prefer business logic to not live in the database)

**Comment by VictorPeralta** [comment, marked as answer]
You can  use the `.or` method in the supabase client

**Reply by soedirgo** [reply]
This is how you'd do it @logikaljay, the caveat is you need to use PostgREST syntax for the values: https://supabase.io/docs/reference/javascript/or https://postgrest.org/en/v8.0/api.html#horizontal-filtering-rows

**Reply by logikaljay** [reply]
Appreciate the response guys, Apologies for the delay but I am still struggling to get this going.

Here is what I have so far:
```js
  let query = supabase
    .from('email_account_messages')
    .select(`
      *
    `)
    .or(`to.cs.[{ "address": "logikal@gmail.com" }]`)
```

This works, when querying for a single address, as soon as I add another email address in to the query:
```js
  let query = supabase
    .from('email_account_messages')
    .select(`
      *
    `)
    .or(`to.cs.[{ "address": "logikal@gmail.com" }, { "address": "someone@else.com" }]`)
```

It fails with the following error:
```
{
  message: '"failed to parse logic tree ((to.cs.[{ "address": "logikal@gmail.com" }, { "address": "someone@else.com" }]))" (line 1, column 47)',
  details: 'unexpected "{" expecting field name (* or [a..z0..9_]), negation operator (not) or logic operator (and, or)'
}
```

I have tried numerous other combinations of square brackets, squiggly brackets (as describe in the postgrest horizontal filtering docs), parentheses etc, but this using the above approach of the array is the closest I have got to getting a working solution.

Any help anyone could provide would be appreciated.

**Reply by steve-chavez** [reply]
@logikaljay Try it like:

```js
  let query = supabase
    .from('email_account_messages')
    .select(`
      *
    `)
    .or(`to.cs.{logikal@gmail.com,someone@else.com"}`)
```

The `cs` value inside `or` is not json despite the curly brackets, it's a postgres array literal:
```sql
select '{logikal@gmail.com,someone@else.com}'::text[];
```

**Reply by logikaljay** [reply]
Hey Steve,

Unfortunately I can't seem to get that to work either.

I have tried it with/without quotes (as your snippet just had a doublequote at the end).
I have tried it with single quotes, double quotes, no quotes, nothing really seems to make a difference.

I am always getting something along the lines of:
```
{
  hint: null,
  message: 'invalid input syntax for type json',
  code: '22P02',
  details: 'Token "logikal" is invalid.'
}
```

I am also failing to see how that approach will actually work, given that `to` is a jsonb field which contains an array of objects:
`{ name: string, email: string }[]`

**Reply by logikaljay** [reply]
The solution (to my problem at least) is as follows:

```js
  let query = supabase
    .from('email_account_messages')
    .select(`
      *
    `)
    .or(`to.cs.[{"address": "email@address.com"}],to.cs.[{"address": "another-email@address.com"}]`)
```

Thanks everyone who devoted the time to help with this!
